proposals/0000-overloaded-record-fields.rst

```diff
@@ -0,0 +1,352 @@
+in the process.
+
+In addition to resolving ambiguous field selectors to a single record
+type using type inference, the proposed change enables definitions
+that are polymorphic over record fields, so for example
+
+::
+
+  \ p -> #personId p + (1 :: Int)
+
+will be given the inferred type
+
+::
+
+  HasField "personId" r Int => r -> Int
+
+and can be applied to any record type with a field ``personId ::
+Int``.
+
+
+Proposed Change
+---------------
+
+HasField class and OverloadedRecordFields extension
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The module ``GHC.Records`` defines the following class:
+
+::
+
+  class HasField (x :: Symbol) r a | x r -> a where
+    fromLabel :: r -> a
+
+When the new extension ``OverloadedRecordFields`` is enabled:
+
+* the existing ``DuplicateRecordFields`` extension is automatically enabled,
+* overloaded label syntax is available, and
+* a label ``#foo`` is translated to
+  ``GHC.Records.fromLabel @"foo" :: HasField "foo" r a => r -> a``.
+
+The third point assumes ``OverloadedLabels`` and ``RebindableSyntax``
+are *disabled* (see `Multiple versions of fromLabel`_ below).
+
+
+Solving HasField constraints
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If the constraint solver encounters a constraint ``HasField x r a``
+where ``r`` is a concrete datatype with a field ``x`` in scope, it
+will automatically solve the constraint using the field selector as
+the dictionary, unifying ``a`` with the type of the field if
+necessary.  This happens irrespective of which extensions are enabled,
+as with other built-in classes with special constraint solving
+behaviour (e.g. ``Coercible``).
+
+For the ``personId`` example above, the end result is rather like
+having an instance
+
+::
+
+  instance HasField "personId" Person Int where
+    fromLabel = personId
+
+except that this instance is not actually generated anywhere, rather
+the constraint is solved directly by the constraint solver.
+
+A field must be in scope for the corresponding ``HasField`` constraint
+to be solved.  This retains the existing representation hiding
+mechanism, whereby a module may choose not to export a field,
+preventing client modules from accessing or updating it directly.
+Thus we cannot actually generate and export ``HasField`` instances
+from defining modules: since there is no mechanism for limiting the
+scope of instances, doing so would expose representation details that
+should be hidden.
+
+Users may define their own instances of ``HasField``, subject to the
+usual rules about overlapping and incoherent instances.  This allows
+"virtual" record fields to be defined for datatypes that do not
+otherwise have them.  For example, an anonymous records library could
+provide ``HasField`` instances and thus be compatible with the
+polymorphic record selectors introduced by ``OverloadedRecordFields``.
+
+
+
+OverloadedLabels extension
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``IsLabel`` class defined in ``GHC.OverloadedLabels`` is changed
```

Comment:
Minor suggestion: perhaps also include the old definition for `IsLabel` inline for ease of comparison in the proposal